File: ctci/chapter_4_trees_graphs/js/4.2-minimalTree.js
const { TreeNode, preOrderTraversal } = require('./tree')

/**
 * Creates a binary search tree with minimal height out of sorted array.
 * @param   {number[]}     arr
 * @returns {TreeNode}
 */
const createMinimalBST = arr => {
  return minimalBSTHelper(arr, 0, arr.length - 1)
}

/**
 * Helper.
 * @param   {number[]}     arr
 * @param   {number}       start
 * @param   {number}       end
 * @returns {TreeNode}
 */
const minimalBSTHelper = (arr, start, end) => {
  if (start > end) return null
  const mid = Math.floor((start + end) / 2)
  const node = new TreeNode(arr[mid])
  node.left = minimalBSTHelper(arr, start, mid - 1)
  node.right = minimalBSTHelper(arr, mid + 1, end)
  return node
}

const arr = [1, 4, 6, 8, 9, 12, 19, 21]
const root = createMinimalBST(arr)
preOrderTraversal(root)
/*
      8
    /   \
  4       12
  / \     /  \
1   6   9    19
              /
            21
*/

File: ctci/chapter_4_trees_graphs/4.2-minimal_tree.py
from typing import List
TreeNode = __import__('tree').TreeNode
pre_order_traversal = __import__('tree').pre_order_traversal


def create_minimal_bst(arr: List[int]) -> TreeNode:
    """
    Creates a binary search tree with minimal height
    out of sorted array.
    """
    return create_minimal_bst_helper(arr, 0, len(arr) - 1)


def create_minimal_bst_helper(arr: List[int], start: int, end: int) -> TreeNode:
    """
    Helper.
    """
    if start > end:
        return None
    mid = (start + end) // 2
    node = TreeNode(arr[mid])
    node.left = create_minimal_bst_helper(arr, start, mid - 1)
    node.right = create_minimal_bst_helper(arr, mid + 1, end)
    return node


if __name__ == "main":
    arr = [1, 4, 6, 8, 9, 12, 19, 21]
    root = create_minimal_bst(arr)
    pre_order_traversal(root)
    r"""
        8
      /   \
    4       12
    / \     /  \
  1   6   9    19
                /
              21
  """
